github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+func (s *RateLimitSuite) TestRateLimitConsecutively() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
+
+	response, err = s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	t1 := time.Now()
+	elapsed = t1.Sub(t0)
+	s.True(elapsed > defaultRateLimitReset, "request took %s, but it should be, at least %s", elapsed, defaultRateLimitReset)
+	s.Contains(s.loggerMock.Next(), "rate limit reached, sleeping until")
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
+
+	response, err = s.transport.RoundTrip(newRequest("/normal"))
+	t2 := time.Now()
+	elapsed = t2.Sub(t1)
+	s.True(elapsed > defaultRateLimitReset, "request took %s, but it should be, at least %s", elapsed, defaultRateLimitReset)
+	s.Contains(s.loggerMock.Next(), "rate limit reached, sleeping until")
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
+
+	response, err = s.transport.RoundTrip(newRequest("/normal"))
+	elapsed = time.Now().Sub(t2)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
+}
+
+// TestRateLimitButWaitInsteadOfRetry ensures that RateLimitTransport does not block requests once the
+// previous RateLimit expired
+func (s *RateLimitSuite) TestRateLimitButWaitInsteadOfRetry() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
+
+	// The next Request is going to wait for more time than the previous RateLimit, so it should not be blocked by RateLimitTransport
+	time.Sleep(defaultRateLimitReset + time.Second)
+
+	t1 := time.Now()
+	response, err = s.transport.RoundTrip(newRequest("/normal"))
+	elapsed = time.Now().Sub(t1)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
+}
+
+// TestAbuse ensures that hitting AbuseRateLimit, causes a wait period
+// it should log the AbuseRateLimit, and wait till the lock expires
+// if the next Request does not hit a RateLimit, it should return the response inmediately
+func (s *RateLimitSuite) TestAbuse() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/abuse_sleep"))
```

Comment:
also here and in other places better check errors as soon as possible because of the same reasoning as described above.

Reply:
globally done by b43b7d2